tools/cibadmin.c

```diff
@@ -597,6 +658,40 @@ main(int argc, char **argv)
     } else if (admin_input_stdin) {
         source = "STDIN";
         input = stdin2xml();
+
+    } else if (acl_eval_how != acl_eval_unused) {
+        username = pcmk__uid2username(geteuid());
+        if (pcmk_acl_required(username)) {
+            if (force_flag) {
+                fprintf(stderr, "The supplied command can provide skewed"
+                                 " result since it is run under user that also"
+                                 " gets guarded per ACLs on their own right."
+                                 " Continuing since --force flag was"
+                                 " provided.");
+
+            } else {
+                fprintf(stderr, "The supplied command can provide skewed"
+                                 " result since it is run under user that also"
+                                 " gets guarded per ACLs in their own right."
+                                 " To accept the risk of such a possible"
+                                 " distortion (without even knowing it at this"
+                                 " time), use the --force flag.");
+                crm_exit(CRM_EX_UNSAFE);
+            }
+
+        }
+        free(username);
+        username = NULL;
+
+        if (cib_user == NULL) {
+            fprintf(stderr, "The supplied command requires -U user specified.");
+            crm_exit(CRM_EX_USAGE);
+        }
+
+        /* we already stopped/warned ACL-controlled users about consequences */
+        acl_cred = cib_user;
+        cib_user = NULL;  /* XXX CRM_DAEMON_USER as it's IPC guarded anyway? */
+        /* XXX should likely differentiate per admin_input_* */
```

Comment:
I think we can drop both these XXX comments (btw he used XXX to mean TODO)

Normally, cib_user is used to run the query as the specified user (instead of root; --user won't do anything if a non-root user runs cibadmin). However, if we did that for ACL highlighting, it would defeat the purpose, since the user couldn't see the parts of the CIB they don't have access to (and thus couldn't show them in red). That's why we set acl_cred instead of cib_user here.

Setting cib_user to CRM_DAEMON_USER instead of NULL would only matter if the CIB_user environment variable were set (if cib_user is NULL, the library will check CIB_user, but if it's set explicitly, the library will use that user, and CRM_DAEMON_USER has permission to see the entire CIB). But CIB_user is not often used, and if it is, I think it's reasonable to leave it as the user's responsibility to figure it out.

I don't see any reason we'd set a different user based on the source of the XML (admin_input_\* indicates whether the input is from a file, pipe, or literal string).